### PR TITLE
Feature/recovery code

### LINKED ISSUE #211: Add support for Recovery Codes
This issue covers implementation of feature called "Recovery Code", which allows activation recovery in case that user loose the mobile phone with an activated, critical application (like our Mobile Key). The feature will be fully documented in our [powerauth-crypto](https://github.com/wultra/powerauth-crypto) repository, once we release it with the rest of the PowerAuth stack. In this ticket, we provide only a brief list of planned features:

- User can obtain the recovery code as a part of activation process (regular or custom). Applications with legacy activations must re-activate the application.
- User can obtain the recovery code via the OOB channel (like via secure postcard, personally at bank branch, etc...). Such code must be confirmed first, before is used for recovery activation.
- Mobile SDK will provide functions to create an activation via the recovery code. The process is quite similar to a regular activation
- Mobile SDK will provide functions to confirm recovery codes delivered via OOB channel.
- Mobile SDK will provide functions to show recovery code created during the activation

### PR DESCRIPTION
**Pull request by hvge**
This PR brings support for "Recovery Codes" feature to mobile SDK:

- All activation types can return recovery data in the activation data result
- Added new functions for getting recovery data on demand (user has to authenticate himself with knowledge factor)
- Added new function for recovery code confirmation on demand.
- Added new functions to `OtpUtils` classes
- On Android, I've moved RESTful model classes directly to the SDK source codes (See #213)
- On Android, we have fixed (@TomasKypta & @kober32) insufficient ProGuard rules (See #216)

You can review this PR on per-commit basis.

Close #211

### CODE REVIEW

File: proj-xcode/Classes/sdk/PowerAuthSDK.h [hunk 1]
```diff
@@ -163,6 +163,23 @@
 													extras:(nullable NSString*)extras
 												  callback:(nonnull void(^)(PA2ActivationResult * _Nullable result, NSError * _Nullable error))callback;
 
+/**
+ Create a new recovery activation with given name, recovery code, puk and additional extras information.
+ 
+ @param name Activation name, for example "John's iPhone".
+ @param recoveryCode Recovery code, obtained either via QR code scanning or by manual entry.
+ @param pu PUK obtainde by manual entry.
```
Comment (kober32) [inline proj-xcode/Classes/sdk/PowerAuthSDK.h]: missing `k` :-)

File: src/PowerAuth/protocol/PrivateTypes.cpp [hunk 2]
```diff
@@ -224,13 +230,99 @@ namespace protocol
 		// Copy external key flag to the SignatureKeys structure
 		pd.sk.usesExternalKey = pd.flags.usesExternalKey;
 		
+		// encrypted recovery data (PD v4)
+		if (reader.currentVersion() >= PD_VERSION_V4) {
+			result = result && reader.readData	(pd.cRecoveryData);
```
[the remark this answers is not in the document]
Comment (hvge) [reply src/PowerAuth/protocol/PrivateTypes.cpp]: It's visually aligned to similar block of code: 
Screen Shot 2019-03-28 at 09 48 41

File: proj-xcode/Classes/sdk/PowerAuthSDK.h (hunk 1 above)
Comment (hvge) [reply proj-xcode/Classes/sdk/PowerAuthSDK.h]: Fixed

File: proj-xcode/Classes/core/PA2Types.mm [hunk 3]
```diff
@@ -325,3 +329,20 @@ void PA2ActivationStep2ParamToStruct(PA2ActivationStep2Param * p2, io::getlime::
...
+
+void PA2RecoveryDataToStruct(PA2RecoveryData * rd, io::getlime::powerAuth::RecoveryData& cpp_rd)
+{
+	cpp_rd.recoveryCode	= cc7::objc::CopyFromNSString(rd.recoveryCode);
+	cpp_rd.puk			= cc7::objc::CopyFromNSString(rd.puk);
+}
+
+PA2RecoveryData * PA2RecoveryDataToObject(const io::getlime::powerAuth::RecoveryData& cpp_rd)
+{
+	if (cpp_rd.isEmpty()) {
```
[the remark this answers is not in the document]
Comment (hvge) [reply proj-xcode/Classes/core/PA2Types.mm]: That never happens in current usage of the function, but I've added a runtime debug assertion for future usage.

File: src/PowerAuth/PublicTypes.cpp [hunk 4]
```diff
@@ -103,6 +103,16 @@ namespace powerAuth
 	}
 
 	
+	//
+	// MARK: - RecoveryData -
+	//
+	
+	bool RecoveryData::isEmpty() const
+	{
+		return recoveryCode.empty() && puk.empty();
```
[the remark this answers is not in the document]
Comment (hvge) [reply src/PowerAuth/PublicTypes.cpp]: Yes that's correct. I'm using this as a simple flag that structure has no data at all. That means that it's still valid - no recovery data available is a correct state. I have then a [full validation routine](https://github.com/wultra/powerauth-mobile-sdk/blob/feature/recovery-code/src/PowerAuth/protocol/PrivateTypes.cpp#L255) in internals, which will not allow to use such combination.

File: proj-xcode/Classes/core/PA2Types.mm (hunk 3 above)
[the remark this answers is not in the document]
Comment (kober32) [reply proj-xcode/Classes/core/PA2Types.mm]: ok

File: src/PowerAuth/PublicTypes.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (kober32) [reply src/PowerAuth/PublicTypes.cpp]: I thought so, ok.

File: proj-android/PowerAuthLibrary/src/main/java/io/getlime/security/powerauth/sdk/PowerAuthSDK.java [hunk 5]
```diff
@@ -58,28 +59,34 @@
...
+import io.getlime.security.powerauth.networking.response.IGetRecoveryDataListener;
 import io.getlime.security.powerauth.networking.response.IValidatePasswordListener;
-import io.getlime.security.powerauth.rest.api.model.entity.ActivationType;
-import io.getlime.security.powerauth.rest.api.model.request.v3.ActivationLayer1Request;
-import io.getlime.security.powerauth.rest.api.model.request.v3.ActivationLayer2Request;
-import io.getlime.security.powerauth.rest.api.model.request.v3.VaultUnlockRequestPayload;
-import io.getlime.security.powerauth.rest.api.model.response.v3.ActivationLayer1Response;
-import io.getlime.security.powerauth.rest.api.model.response.v3.ActivationLayer2Response;
-import io.getlime.security.powerauth.rest.api.model.response.v3.VaultUnlockResponsePayload;
+import io.getlime.security.powerauth.networking.model.entity.ActivationType;
```
Comment (TomasKypta) [inline proj-android/PowerAuthLibrary/src/main/java/io/getlime/security/powerauth/sdk/PowerAuthSDK.java]: It's good to "Code -> Optimize Imports" in Android Studio. It'll sort import lexicographically.
The same in other files where the package name has been changed.

Comment (hvge) [reply proj-android/PowerAuthLibrary/src/main/java/io/getlime/security/powerauth/sdk/PowerAuthSDK.java]: Fixed

**Pull request merged**